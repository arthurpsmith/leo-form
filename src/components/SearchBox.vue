<template>
    <v-text-field
      prepend-inner-icon="mdi-magnify"
      placeholder="Search"
      :class="{'expanding-search': true, 'closed': searchClosed   }"
      :value="searchText"
      @focus="toggle"
      @blur="toggle"
      @input="onChange"
      single-line
      clearable
      solo-inverted
      hide-details
      filled
      outlined
      dense
    ></v-text-field>
</template>

<script lang="ts">
import { defineComponent, computed } from '@vue/composition-api'
import { useSearchService } from '@/statemachines/search.machine'
import { StateValue } from 'xstate'

export default defineComponent({
  name: 'SearchButton',
  setup () {
    const searchBoxToggleMachine = useSearchService()
    const searchText = computed(() => {
      // useActor type bug https://github.com/statelyai/xstate/issues/2727
      /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      return searchBoxToggleMachine.state.value.context.text
    })
    const toggle = () => {
      /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      searchBoxToggleMachine.send('TOGGLE')
    }
    const onChange = (e: string) => {
      searchBoxToggleMachine.send({
        /*
      Ignore upstream type definition bug
      https://github.com/statelyai/xstate/issues/2727
       */
        /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
        // @ts-ignore
        type: 'REPORT_TEXT_CHANGE',
        /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
        // @ts-ignore
        text: e
      })
    }
    return {
      searchBoxToggleMachine, searchText, onChange, toggle
    }
  },
  computed: {
    searchBoxState (): StateValue {
      return this.searchBoxToggleMachine.state.value.value
    },
    isShown (): boolean {
      return this.searchText.length > 0 || this.searchBoxState === 'shown'
    },
    searchClosed (): boolean {
      return this.searchBoxState === 'hidden'
    }
  }
})
</script>

<style lang="sass">
  .v-input.expanding-search
    transition: max-width 0.3s ease-out
    max-width: 200px
    &.closed
      transition: max-width 0.3s ease-in
      max-width: 45px
</style>
