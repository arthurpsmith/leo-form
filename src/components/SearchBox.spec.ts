import { createLocalVue, mount, Wrapper } from '@vue/test-utils'
import SearchBox from '@/components/SearchBox.vue'
import Vuetify from 'vuetify'
import { getSearchService, searchSymbol, useSearchService } from '@/statemachines/search.machine'
import CompositionApi from '@vue/composition-api'
import { useActor } from 'xstate-vue2'

const toggleSearchBox = async (wrapper: Wrapper<any>) => {
  const input = wrapper.find('input')
  await input.trigger('focus')
}

const updateSearchText = async (wrapper: Wrapper<any>, searchText: string) => {
  const input = wrapper.find('input')
  await toggleSearchBox(wrapper)
  await input.setValue(searchText)
}

describe('SearchBox.vue', () => {
  const localVue = createLocalVue()
  localVue.use(CompositionApi)
  let vuetify: Vuetify
  let searchServiceToProvide: ReturnType<typeof getSearchService>
  let provideWithExportedSymbol: Record<string, unknown>
  let searchService

  function mountFunction (options = {}) {
    return mount(SearchBox, {
      localVue,
      vuetify,
      provide: provideWithExportedSymbol,
      ...options
    })
  }

  function mountAttachedToDOM () {
    /*
    When triggering 'focus' with jsdom v16.4.0+, the component must be mounted with the
    attachTo option, see https://vue-test-utils.vuejs.org/api/wrapper/trigger.html
     */
    const div = document.createElement('div')
    div.id = 'root'
    document.body.appendChild(div)
    const wrapper = mountFunction({
      attachTo: '#root'
    })
    return wrapper
  }
  beforeEach(() => {
    vuetify = new Vuetify()
    searchServiceToProvide = getSearchService()
    provideWithExportedSymbol = {
      [searchSymbol as symbol]: searchServiceToProvide
    }
    searchService = useActor(searchServiceToProvide)
  })
  it('should work', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
  })
  it('Has the right icon', () => {
    const wrapper = mountFunction()
    expect(wrapper.find('i.v-icon').exists()).toBe(true)
  })
  it('Has the correct initial classnames', () => {
    const wrapper = mountFunction()
    expect(wrapper.classes()).toEqual(expect.arrayContaining(['expanding-search', 'closed']))
  })
  it('Has the correct initial (blank) search text', () => {
    const wrapper = mountFunction()
    /*
       VTU wrapper is typed as a generic Vue instance and can't infer the types from the component
       when it's an SFC, so we must declare the component - wrapper.vm - as any
       https://github.com/vuejs/vue-test-utils/issues/255
     */
    expect((wrapper.vm as any).searchText).toEqual('')
  })
  it('Has the correct classnames on click to open', async () => {
    const wrapper = mountAttachedToDOM()
    await toggleSearchBox(wrapper)
    expect(wrapper.classes()).not.toContain('closed')
    /*
    https://vue-test-utils.vuejs.org/api/options.html#attachto
    When attaching to the DOM, you should call wrapper.destroy() at the end of your test to remove
    the rendered elements from the document and destroy the component instance.
     */
    wrapper.destroy()
  })
  it('Updates the search text on input', async () => {
    const wrapper = mountAttachedToDOM()
    const newSearchText = 'Some new search text'
    await updateSearchText(wrapper, newSearchText)
    expect((wrapper.vm as any).searchText).toEqual(newSearchText)
    wrapper.destroy()
  })
  it('Applies the closed class on blur, when search text is empty', async () => {
    const wrapper = mountAttachedToDOM()
    await toggleSearchBox(wrapper)
    const input = wrapper.find('input')
    await input.trigger('blur')
    expect(wrapper.classes()).toContain('closed')
  })
  it('Does not apply the closed class on blur, when the search text is not empty', async () => {
    const wrapper = mountAttachedToDOM()
    const newSearchText = 'Some other new search text in Korean 한국어로 된 다른 새로운 검색 텍스트'
    await updateSearchText(wrapper, newSearchText)
    expect((wrapper.vm as any).searchText).toEqual(newSearchText)
    const input = wrapper.find('input')
    input.trigger('blur')
    expect(wrapper.classes()).not.toContain('closed')
  })
})
