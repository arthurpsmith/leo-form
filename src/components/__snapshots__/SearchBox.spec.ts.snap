// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox.vue should work 1`] = `
<div class="v-input v-input--hide-details v-input--dense theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--solo-inverted v-text-field--filled v-text-field--enclosed v-text-field--outlined v-text-field--placeholder expanding-search closed">
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-input__prepend-inner">
        <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-magnify theme--light"></i></div>
      </div>
      <fieldset aria-hidden="true">
        <legend><span>â€‹</span></legend>
      </fieldset>
      <div class="v-text-field__slot"><input id="input-4" placeholder="Search" type="text"></div>
      <div class="v-input__append-inner">
        <div></div>
      </div>
    </div>
  </div>
</div>
`;
